### ISSUE TITLE
Implement diyHue

### ISSUE THREAD

**Issue by pete111** [Aircoookie/WLED#40]
Hi @Aircoookie , because I don't have Philips Hue bridge, I am trying it with this emulator https://github.com/mariusmotea/diyHue. Philips Hue mobile app successfully found this bridge, but when I save bridge IP in Wled, no wled device is added to Hue bridge.  I also initiated http://bridgeIP/hue/linkbutton?username=Hue&password=Hue&action=Activate (on diyHue) but I am constantly getting Hue status: Inactive on Wled. Could you please try it, too? Probably I should buy one of that expensive Philips bridges :)

**Comment by Aircoookie** [comment]
Hey,

it's not your Hue emulators fault. WLED can't emulate a Hue lamp, all it does is poll the state of an official hue lamp every few seconds and set itself to the same color. Try adding a virtual hue light in your hue emulator and then, in wled sync settings, tick the checkbox after "poll hue light 1 every X ms". You can thrn change the wled light by changing the virtual hue light in the hue app. However I don't think this would make sense if you just use WLED, it is intended to sync WLED to some other real lights :)

**Comment by Aircoookie** [comment]
I know, I'd love that feature! I'll add it as soon as the hue protocol supports adding "non-Zigbee" lights! (which it currently doesn't, unfortunately)

**Comment by pete111** [comment]
great 👍 
I currently tested color cloning from hue bulb (emulated). It mostly works, but from time to time esp restarts.

**Comment by Aircoookie** [comment]
Yeah, I'll admit it is not the most stable. It helps a bit if you increase the polling interval to 2000ms or something, but of course this means more latency. The ESP doesn't like sending requests all the time somehow. But currently this is the only possible way to sync to hue lights :)

**Comment by juanesf** [comment]
Hello Aircoookie and thanks for your excellent work I also use the @mariusmotea emulator and you could implement your sketch to add direct support to the emulator

https://github.com/mariusmotea/diyHue/blob/master/Lights/Arduino/WS2812BHueStrip/WS2812BHueStrip.ino

**Comment by nightflasher** [comment]
Hi, I know this issue is a bit old, but I've found an older project I've used till last weekend I've found yours. It's the esp hue emulator which is controlled by any third party app for the hue system (the ones that I've used so far worked quite well). Please have a look at this and eventually you could implement this?
https://github.com/probonopd/ESP8266HueEmulator

**Comment by Aircoookie** [comment]
In fact I'm already using similar code that was inspired by that project for the Alexa function! Integrating the emulator into WLED would add compatibility with many hue clients, which would be super awesome, but it is quite complex and not a library, so I'd need to do some restructuring to make it compatible and I'm afraid it might have a high performance impact if I add too many new interfaces. Still, I'll keep this in mind and will consider implementing it sometime soon!

**Comment by juanesf** [comment]
hello again @Aircoookie, @mariusmotea has several sketch to emulate huestrip and huelamps for the diyHue emulator which also has support for hue entertainment.
the next sketch has the option to define the amount of lights to emulate by a led strip, and also define from which led the following light begins:
https://github.com/diyhue/Lights/blob/master/Arduino/WS2812BHueRing/WS2812BHueRing.ino

while the following can define the amount of lights to emulate per LED strip:
https://github.com/diyhue/Lights/blob/master/Arduino/WS2812BHueRing/WS2812BHueRing.ino

with those sketchs you can have direct support from diyHue which can be controlled by an Original Hue Bridge and extend the lights by wifi without the need to add ZigBee. The emulator can be installed on Raspberry, linux, Docker, OpenWrt and I think I saw on YouTube a video where it was running on Windows. Any additional info can enter the chat in Slack:
https://slackinvite.squishedmooo.com/

**Comment by juanesf** [comment]
or perhaps add only the hue entertainment protocol

**Comment by juanesf** [comment]
void entertainment() {
  uint8_t packetSize = Udp.parsePacket();
  if (packetSize) {
    if (!entertainment_run) {
      entertainment_run = true;
    }
    lastEPMillis = millis();
    Udp.read(packetBuffer, packetSize);
    for (uint8_t i = 0; i < packetSize / 4; i++) {
      rgb[packetBuffer[i * 4]][0] = packetBuffer[i * 4 + 1];
      rgb[packetBuffer[i * 4]][1] = packetBuffer[i * 4 + 2];
      rgb[packetBuffer[i * 4]][2] = packetBuffer[i * 4 + 3];
    }
    for (uint8_t light = 0; light < lightsCount; light++) {
      if (lightsCount > 1) {
        if (light == 0) {
          for (uint8_t pixel = 0; pixel < lightLedsCount + transitionLeds / 2; pixel++) {
            if (pixel < lightLedsCount - transitionLeds / 2) {
              strip.SetPixelColor(pixel, convInt(rgb[light]));
            } else {
              //strip.SetPixelColor(pixel, RgbColor(rgb[light][0], rgb[light][1], rgb[light][2]));
              strip.SetPixelColor(pixel, blendingEntert(rgb[0], rgb[1], pixel + 1 - (lightLedsCount - transitionLeds / 2 )));
            }
          }
        } else if (light == lightsCount - 1) {
          for (uint8_t pixel = 0; pixel < lightLedsCount - transitionLeds / 2 ; pixel++) {
            strip.SetPixelColor(pixel + transitionLeds / 2 + lightLedsCount * light, convInt(rgb[light]));
          }
        } else {
          for (uint8_t pixel = 0; pixel < lightLedsCount; pixel++) {
            if (pixel < lightLedsCount - transitionLeds) {
              strip.SetPixelColor(pixel + transitionLeds / 2 + lightLedsCount * light, convInt(rgb[light]));
            } else {
              strip.SetPixelColor(pixel + transitionLeds / 2 + lightLedsCount * light, blendingEntert(rgb[light], rgb[light + 1], pixel - (lightLedsCount - transitionLeds ) + 1));
            }
          }
        }
      } else {
        strip.ClearTo(RgbColor(rgb[0][0], rgb[0][1], rgb[0][2]), 0, lightLedsCount - 1);
      }
    }
    strip.Show();
  }
}

**Comment by juanesf** [comment]
This is the link to the last state that worked for diyHue which can serve as something

https://github.com/diyhue/Lights/blob/master/Arduino/wled00/wled06_usermod.ino

**Comment by juanesf** [comment]
And @alexyao2015 has added a protocol for ESPHome that can serve as a guide, regards

https://github.com/diyhue/diyHue/commits/master/BridgeEmulator/protocols/esphome.py

**Comment by snickers2k** [comment]
+1
maybe this helps: https://github.com/linuzer/Arduino-HueAPI

**Comment by bravery10** [comment]
Why can I sync with Hue Bridge but not with Hue Emulator, I clicked activate on Huelinkbuton page and entered the correct number of lights and checked all box.

**Comment by juanesf** [comment]
This thread should be closed since diyHue makes discovery directly, it is not necessary to make any changes in wled

**Comment by juanesf** [comment]
What you should consider is that diyHue is in a protocol transition and the branch that works best is beta but I recommend using docker

**Comment by bravery10** [comment]
thank you! i am using docker on raspberry 4B, i have 1 Hue brigde v2 and 1 Hue emulator, i want use wled to clone philip hue light or diyhue light, wled is being used as a ambilight.

**Comment by Aircoookie** [comment]
I think this could be implemented well as a usermod, adapted from this sketch: https://github.com/diyhue/Lights/blob/master/ESP8266/Hue_Gradient_Lightstrip_SK6812/Hue_Gradient_Lightstrip_SK6812.ino

**Comment by juanesf** [comment]
The support in diyhue is complete, it is not necessary. There is already a wled protocol in diyhue.

https://github.com/diyhue/diyHue/blob/master/BridgeEmulator/lights/protocols/wled.py
